File: src/advent_of_code_2015/Day10/main.py
def look_and_say(s: str) -> str:
    result = []
    count = 1

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            count += 1
        else:
            result.append(str(count))
            result.append(s[i - 1])
            count = 1

    # last group
    result.append(str(count))
    result.append(s[-1])

    return "".join(result)


def main(input_str: str, times: int = 40) -> int:
    current = input_str

    for _ in range(times):
        current = look_and_say(current)

    return len(current)


print(main("1113222113", 50))

# def main (input: str) -> int:
#     map = {}
#     result = 0
#     for _ in 40:
#         for i in range(len(input)):
#             key = int(input[i])
#             map[key] = map.get(key, 0) + 1
#             if i > 0 and key != int(input[i - 1]):
#                 previous_key = int(input[i - 1])
    
#                 if result == 0:
#                     result += map[previous_key] * 10
#                     result += previous_key
#                 else:
#                     result *= 10
#                     result += map[previous_key]
#                     result *= 10
#                     result += previous_key
#                 del map[previous_key]
    
#     #post process
#     for key, value in map.items():
#         result *= 10
#         result += value
#         result *= 10
#         result += key

#     return result


# print(main("111221"))

File: src/advent_of_code_2015/Day10/main.cs
using System.Text;

class Program
{
    static void Main()
    {
        string currentString = "1113222113";
        for (int i = 0; i < 40; i++)
        {
            currentString = LookAndSay(currentString);
        }

        Console.WriteLine(currentString.Length);
    }

    private static string LookAndSay(string input)
    {
        StringBuilder sb = new StringBuilder();
        int count = 1;
        
        for (int j = 1; j < input.Length; j++)
        {
            if (input[j] != input[j - 1])
            {
                sb.Append(count);
                sb.Append(input[j - 1]);
                count = 1;
                
            } else
            {
                count++;
            }
        }

        //post processing
        sb.Append(count);
        sb.Append(input[input.Length - 1]);
        return sb.ToString();
    }
}